### PR TITLE
CP-14655: Fix activity swap-title token parsing for recurring fills & approvals

### PR DESCRIPTION
**Pull request by ruijia1in**
iOS: 9135
Android: 9136

## Description

**Ticket: [CP-14655]**

Please provide:

The Activity tab mislabeled two kinds of Markr transactions:

- Recurring-swap fills (USDC→USDT) showed "0.05 AVAX swapped for 0.99 USDT" because getSwapTitle took inputTokens[0], and the native schedule-fee leg happened to precede the ERC-20 principal. Extract the leg selection into a pure selectSwapTokens() helper that prefers the non-native input leg (the native leg is a protocol/gas fee, not the traded principal); a genuine native-input swap still resolves correctly.

- Single-leg contract calls with no output (e.g. an ERC-20 approval that the indexer returns as txType UNKNOWN with no method) showed "10 USDC swapped". These are indistinguishable from a cross-chain swap input by their token legs, so only label "swapped" when the indexer classifies the tx as SWAP; otherwise fall back to "<amount> <symbol> Contract Call".

Adds unit tests for selectSwapTokens covering the recurring-fill fee-leg case, native-input swaps, plain ERC-20 swaps, cross-chain input-only, and the no-user-leg case.

### Also included: Recurring schedules "In a few minutes" label ([CP-14659])

On the Recurring Schedules screen, a `nextExecutionAt` in the past means the order was just resumed (or the relayer hasn't caught up yet) — it will fire on the next relayer run rather than at the stale scheduled timestamp. Showing the past timestamp was misleading, so we now render a soft "In a few minutes" label instead. Matches the CP-14659 web behavior.

### Also included: Recurring-swap Android layout fixes (Fabric)

On Android (new architecture / Fabric), the recurring-swap UI clipped and overlapped content:

- Schedule cards on the Manage screen cut off the "Next swap scheduled" row.
- The recurring details rows clipped when a row expanded.
- On the Swap screen, the "Swap rate is for the first swap…" disclaimer overlapped the swap card above and the toggle below.

Root cause: a Reanimated `LinearTransition` view with `overflow: 'hidden'` settles at a stale (too-short) height when its children mount/grow, and sibling views don't reliably reflow when a new view mounts between them.

Fix: move `overflow: 'hidden'` off the layout-animated views onto static inner clip wrappers (`GroupList`, `ScheduleCard`, `RecurringDetailsRows`) so the clip measures the real content height; the rounded surface + caller styles stay on the animated node so existing `GroupList` callers (StakeDetail radius override, Collectible absolute positioning) are unaffected. On the Swap screen, the recurring section is wrapped in a container keyed on `isRecurring` so toggling forces a clean layout pass (same CP-14600 pattern), and the disclaimer text uses `flex: 1` for reliable wrapping.

### Also included: Android form-sheet scroll dismisses instead of scrolling

On Android, scrolling down inside a form-sheet modal (e.g. the Swap screen) dismissed the whole sheet instead of scrolling the content. `ScrollScreen`'s scroll containers didn't opt into Android nested scrolling, so Material's `BottomSheetBehavior` captured the vertical swipe as drag-to-dismiss. Set `nestedScrollEnabled={Platform.OS === 'android'}` on both the keyboard-aware and plain scroll paths so content scrolls normally and the sheet only dismisses when dragging down from the top (matches the CP-14372 pattern in StakeSearchScreen / ListScreenV2). Fix lives in `ScrollScreen`, so every modal built on it benefits.

### Also included: Recurring schedules empty state centered

The empty (and empty-error) placeholder on the Manage screen was pinned near the top via `paddingTop`. It now lets the scroll content grow to fill the modal and centers the placeholder in the remaining space, matching the loading branch's centering pattern.

> Verified via tsc / eslint / unit tests (129 recurring-swap tests pass). Still needs on-device Android confirmation.

## Screenshots/Videos

IMG_3030

https://github.com/user-attachments/assets/1893a631-ba7c-444d-85d5-cbb8a3976a23

Screenshot_20260702_141611_Core Internal

## Testing

**Dev Testing (if applicable)**

- Provide steps to test the happy path of your feature
- Provide steps to test edge cases and error states
- Trigger a build on bitrise and reference it here
- Move the ticket into the "Testing" column on Jira

**QA Testing (if applicable)**

- Provide instructions for QA to test this feature thoroughly
- State expected behavior / acceptance criteria

## Checklist

Please check all that apply (if applicable)

- [ ] I have performed a self-review of my code
- [ ] I have verified the code works
- [ ] I have included screenshots / videos of android and ios
- [ ] I have added testing steps
- [ ] I have added/updated necessary unit tests
- [ ] I have updated the documentation

[CP-14655]: https://ava-labs.atlassian.net/browse/CP-14655?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CP-14659]: https://ava-labs.atlassian.net/browse/CP-14659

### CODE REVIEW

File: packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx [hunk 1]
```diff
@@ -89,63 +90,71 @@ export const TokenActivityListItemTitle = ({
...
             ' swapped'
           ]
         }
+        // Keep the leg amount + symbol so the row still conveys what moved
+        // (e.g. "10 USDC Contract Call"), matching the "10 USDC sent" style
+        // of the other single-leg labels.
+        return [
+          renderAmount(inputToken.amount),
+          ' ',
+          inputToken.symbol,
```
Comment (atn4z7) [inline packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx]: This renders `inputToken.symbol` raw, but the comment says it matches the "10 USDC sent" style of the other single-leg labels — those actually use `s1` from `getIOTokenAmountAndSymbol()`, which falls back to `token.type` when `symbol` is empty. Since `selectSwapTokens` only returns a native leg when it's the *only* input, the trigger here is a native-input contract call (typed non-SWAP) with no output leg, and per utils.ts (`resolvePaymentSymbol`) "Glacier often returns empty symbol for native assets." That renders as `0.05  Contract Call` (symbol dropped, double space) instead of `0.05 AVAX …`. Minor/cosmetic, but consider routing the symbol through the network-token fallback so it matches the parity goal stated in the comment.

File: packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx [hunk 2]
```diff
@@ -89,63 +90,71 @@ export const TokenActivityListItemTitle = ({
...
+      //   • a cross-chain swap whose output token lands on another chain
+      //   • an unclassified contract call with a single leg — e.g. an ERC-20
+      //     approval, whose Approval event the indexer surfaces as a
+      //     user→spender token with no return.
+      // The indexer classifies genuine swaps as `txType === SWAP`, so only
+      // trust the "swapped" label there. Otherwise (`UNKNOWN` etc.) `method`
+      // and `txType` give us nothing to prove a swap or an approval, so fall
+      // back to a neutral "Contract Call" rather than fabricating either.
+      if (inputToken) {
+        if (transaction.txType === TransactionType.SWAP) {
```
Comment (atn4z7) [inline packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx:155]: Not a defect — just flagging the tradeoff: the old input-only branch explicitly targeted cross-chain swaps whose output lands on another chain, and those now fall to "Contract Call" unless typed `SWAP`. In practice the risk looks low (such txs are typically `SWAP`, handled above, or `BRIDGE`, handled in the switch, so they don't reach here), but worth a sanity check against how the indexer classifies Markr cross-chain fills. If they carry `txType === SWAP`, no change needed.

**Review by atn4z7 — COMMENTED** [review]
Nice PR.

File: packages/core-mobile/app/new/features/activity/utils.ts [hunk 3]
```diff
@@ -239,3 +239,34 @@ export const getNftLabel = (nftToken: TxToken | undefined): string => {
...
+    token => token.from?.address === userAddress
+  )
+  const outputTokens = tokens.filter(token => token.to?.address === userAddress)
+
+  const inputToken =
+    inputTokens.find(token => token.type !== TokenType.NATIVE) ?? inputTokens[0]
+  const outputToken = outputTokens[0]
+
+  return { inputToken, outputToken }
+}
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/activity/utils.ts:282]: Fixed in `d1e1eb9b3` (Copilot Autofix): `selectSwapTokens` now early-returns when `userAddress` is missing and compares `from`/`to` addresses via `.toLowerCase()`, matching `findPaymentToken`. The `undefined === undefined` mislabeling is no longer possible.

File: packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx (hunk 1 above)
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx]: Good catch — fixed in `8ca80f77c`. Both legs now go through `resolvePaymentSymbol(token, network?.networkToken.symbol)`, which is the same fallback chain `s1` uses (token.symbol → network token symbol when NATIVE → token.type). A native-input contract call now renders `0.05 AVAX Contract Call` instead of `0.05  Contract Call`.

File: packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx [hunk 4]
```diff
@@ -89,63 +90,71 @@ export const TokenActivityListItemTitle = ({
...
+        return [
+          renderAmount(inputToken.amount),
+          ' ',
+          inputToken.symbol,
+          ' swapped for ',
+          renderAmount(outputToken.amount),
+          ' ',
+          outputToken.symbol
+        ]
       }
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx:142]: Fixed in `8ca80f77c`. Symbol resolution is now hoisted above the branches so the both-legs case uses `resolvePaymentSymbol` too (consolidated with the `6d3d12dcc` autofix, which had added it only to this branch). All three branches — both-legs, input-only "swapped", and "Contract Call" — now share the resolved `inputSymbol`/`outputSymbol`.

File: packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx [hunk 5]
```diff
@@ -89,63 +90,71 @@ export const TokenActivityListItemTitle = ({
...
+      //     user→spender token with no return.
+      // The indexer classifies genuine swaps as `txType === SWAP`, so only
+      // trust the "swapped" label there. Otherwise (`UNKNOWN` etc.) `method`
+      // and `txType` give us nothing to prove a swap or an approval, so fall
+      // back to a neutral "Contract Call" rather than fabricating either.
+      if (inputToken) {
+        if (transaction.txType === TransactionType.SWAP) {
           return [
             renderAmount(inputToken.amount),
             ' ',
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx]: Fixed in `8ca80f77c` — the input-only branch now uses the hoisted `inputSymbol` (via `resolvePaymentSymbol`) in both the "swapped" and "Contract Call" returns, so a native leg with an empty Glacier symbol no longer renders blank.

File: packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx (hunk 2 above)
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx:155]: Checked this against a real Glacier dump for the affected Markr account (74 txs):

- **All txs are single-chain (C-Chain, 43114)** — Markr recurring swaps are same-chain DEX swaps, so there are no cross-chain fills in the actual data.
- **Every Markr fill comes back as `txType: "UNKNOWN"`, `method: null`** — none are typed `SWAP`. Same-chain fills still label correctly because they carry both an input and output leg (→ the `inputToken && outputToken` branch), independent of txType.
- The single-leg input-only `UNKNOWN` txs in the dump are protocol deposits/approvals (e.g. user → Markr contract, no return leg), for which "Contract Call" is the correct neutral label.

**Bridge is provably safe:** `convertTransaction` forces `txType = BRIDGE` whenever `bridgeAnalysis.isBridgeTx`, and the `BRIDGE` case in the switch handles it before `getSwapTitle` is ever reached, so a bridge tx can't fall through to "Contract Call".

**The residual gap** (a cross-chain swap the indexer leaves `UNKNOWN`, output on another chain → input-only) is structurally indistinguishable from an approval at this layer: one `user → contract` leg, no output, no method. `analyzeTx` can't help — it's a pass-through to `@avalabs/bridge-unified`, whose `AnalyzeTxResult` only recognizes known *bridge* protocols (cctp/ictt/lombard/…), not cross-chain swaps, and its input doesn't even carry per-token destination-chain info. So there's no reliable positive signal to add here without an upstream indexer/SDK change.

Net: no code change for this branch — `txType === SWAP` already relabels correctly whenever the indexer does classify it. Left the explanatory comment in place documenting the tradeoff.

File: packages/core-mobile/app/new/features/recurringSwap/screens/RecurringSchedulesScreen.tsx [hunk 6]
```diff
@@ -281,6 +281,18 @@ function ScheduleCard({
     if (s.nextExecutionAt === null || s.nextExecutionAt === undefined) {
       return '—'
     }
+    // A `nextExecutionAt` in the past means the order was just resumed (or the
+    // relayer hasn't caught up yet): it will be triggered on the next relayer
+    // run — usually within a minute. We don't know the exact time, so showing
+    // the stale scheduled timestamp would be misleading. Show a soft label
+    // instead. (Matches the CP-14659 web behavior.)
+    if (s.nextExecutionAt * 1000 <= Date.now()) {
+      return (
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/recurringSwap/screens/RecurringSchedulesScreen.tsx:300]: Fixed in `83643ce37`. Extracted a per-render `isNextExecutionStale` boolean:

```ts
const isNextExecutionStale =
  s.nextExecutionAt !== null &&
  s.nextExecutionAt !== undefined &&
  s.nextExecutionAt * 1000 <= Date.now()
```

and added it to the `nextSwapValue` memo deps, so once the scheduled time passes the memo recomputes on the next render and flips to "In a few minutes" instead of caching the timestamp.

Note this makes the label update on the *next re-render* after the threshold — it doesn't schedule a timer to force a re-render exactly at `nextExecutionAt`. That matches the intended behavior here (the list re-renders on focus/refetch); a live self-updating countdown would be a separate change if we ever want it.

File: packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx [hunk 7]
```diff
@@ -89,63 +90,79 @@ export const TokenActivityListItemTitle = ({
...
+        // Keep the leg amount + symbol so the row still conveys what moved
+        // (e.g. "10 USDC Contract Call"), matching the "10 USDC sent" style
+        // of the other single-leg labels.
+        return [
+          renderAmount(inputToken.amount),
+          ' ',
+          inputToken.symbol,
+          ' Contract Call'
+        ]
       }
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx:167]: Fixed in `8ca80f77c`. Symbol resolution is now hoisted above all branches and both the "swapped" and "Contract Call" returns use the resolved `inputSymbol` (via `resolvePaymentSymbol(inputToken, network?.networkToken.symbol)`), so a native leg with an empty Glacier symbol no longer renders blank.

**Review by atn4z7 — DISMISSED** [review]

**Review by atn4z7 — DISMISSED** [review]

**Review by atn4z7 — DISMISSED** [review]

**Review by atn4z7 — DISMISSED** [review]
reviewed and dev tested

**Review by atn4z7 — DISMISSED** [review]

File: packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx [hunk 8]
```diff
@@ -89,63 +90,79 @@ export const TokenActivityListItemTitle = ({
...
+      //   • a cross-chain swap whose output token lands on another chain
+      //   • an unclassified contract call with a single leg — e.g. an ERC-20
+      //     approval, whose Approval event the indexer surfaces as a
+      //     user→spender token with no return.
+      // The indexer classifies genuine swaps as `txType === SWAP`, so only
+      // trust the "swapped" label there. Otherwise (`UNKNOWN` etc.) `method`
+      // and `txType` give us nothing to prove a swap or an approval, so fall
+      // back to a neutral "Contract Call" rather than fabricating either.
+      if (inputToken) {
+        if (transaction.txType === TransactionType.SWAP) {
```
Comment (B0Y3R-AVA) [inline packages/core-mobile/app/new/features/portfolio/assets/components/TokenActivityListItemTitle.tsx:155]: This `txType === TransactionType.SWAP` guard can only ever be `true` when `getSwapTitle` is entered via the `case TransactionType.SWAP` arm. When it's reached through the `default` → `isContractCall` heuristics (1/2/3-token contract calls), `tx.txType` is by definition **not** `SWAP` (it would have returned from the SWAP case already), so every input-only resolution on that path now falls through to `"Contract Call"`.

That population overlaps with txs the rest of the screen still treats as swaps: `fixUnknownTxType()` normalizes these same `UNKNOWN` leg shapes to `SWAP` for the row **icon**, and the Swap filter matches on `fixUnknownTxType(...) === SWAP`. So an input-only router/cross-chain swap now renders `"10 USDC Contract Call"` while still carrying a swap icon and appearing under the **Swap** filter — previously it was `"10 USDC swapped"`.

Is relabeling these as "Contract Call" intentional? If the goal is only to catch ERC-20 approvals, note approvals are also `UNKNOWN` and also normalize to `SWAP`, so raw `txType` can't separate them from genuine single-leg swaps — gating on `fixUnknownTxType(tx, isFromAccount) === SWAP` instead would keep normalized swaps as "swapped" and stay consistent with the icon/filter.

**Review by atn4z7 — APPROVED** [review]

**Review by B0Y3R-AVA — APPROVED** [review]
